pkg/trie/triedb/proof/generate.go

```diff
@@ -0,0 +1,450 @@
+	childIndex int
+	// children contains the child references to use in constructing the proof nodes.
+	children [codec.ChildrenCapacity]triedb.ChildReference
+	// outputIndex is the index into the proof vector that the encoding of this entry should be placed at.
+	outputIndex *int
+}
+
+func newStackEntry(
+	prefix []byte,
+	nodeData []byte,
+	nodeHash *[]byte,
+	outputIndex *int) (*stackEntry, error) {
+	node, err := codec.Decode(bytes.NewReader(nodeData))
+	if err != nil {
+		return nil, err
+	}
+
+	return &stackEntry{
+		prefix:      prefix,
+		node:        node,
+		encodedNode: nodeData,
+		nodeHash:    nodeHash,
+		omitValue:   false,
+		childIndex:  0,
+		children:    [codec.ChildrenCapacity]triedb.ChildReference{},
+		outputIndex: outputIndex,
+	}, nil
+}
+
+func (e *stackEntry) encodeNode() ([]byte, error) {
+	switch n := e.node.(type) {
+	case codec.Empty:
+		return e.encodedNode, nil
+	case codec.Leaf:
+		if !e.omitValue {
+			return e.encodedNode, nil
+		}
+
+		encodingBuffer := bytes.NewBuffer(nil)
+		err := triedb.NewEncodedLeaf(e.node.GetPartialKey(), codec.NewInlineValue([]byte{}), encodingBuffer)
+		if err != nil {
+			return nil, err
+		}
+
+		return encodingBuffer.Bytes(), nil
+	case codec.Branch:
+		var value codec.EncodedValue
+		if !e.omitValue {
+			value = n.Value
+		}
+		e.completBranchChildren(n.Children, e.childIndex)
+		encodingBuffer := bytes.NewBuffer(nil)
+		err := triedb.NewEncodedBranch(e.node.GetPartialKey(), e.children, value, encodingBuffer)
+		if err != nil {
+			return nil, err
+		}
+		return encodingBuffer.Bytes(), nil
+	default:
+		panic("unreachable")
+	}
+}
+
+func (e *stackEntry) setChild(encodedChild []byte) {
+	var childRef triedb.ChildReference
+	switch n := e.node.(type) {
+	case codec.Branch:
+		if e.childIndex >= codec.ChildrenCapacity {
+			panic("child index out of bounds")
+		}
+		if n.Children[e.childIndex] != nil {
+			childRef = e.replaceChildRef(encodedChild, n.Children[e.childIndex])
+		}
+	default:
+		panic("Empty and leaf nodes have no children, we cannot descend into")
+	}
+	e.children[e.childIndex] = childRef
+	e.childIndex++
+}
+
+func (e *stackEntry) completBranchChildren(
+	childHandles [codec.ChildrenCapacity]codec.MerkleValue,
+	childIndex int,
+) {
+	for i := childIndex; i < codec.ChildrenCapacity; i++ {
+		switch n := childHandles[i].(type) {
+		case codec.InlineNode:
+			e.children[i] = triedb.NewInlineChildReference(n.Data)
+		case codec.HashedNode:
+			e.children[i] = triedb.NewHashChildReference(common.Hash(n.Data))
+		}
+	}
+}
+
+func (e *stackEntry) replaceChildRef(encodedChild []byte, child codec.MerkleValue) triedb.ChildReference {
+	switch child.(type) {
+	case codec.HashedNode:
+		return triedb.NewInlineChildReference(nil)
+	case codec.InlineNode:
+		return triedb.NewInlineChildReference(encodedChild)
+	default:
+		panic("unreachable")
+	}
+}
+
+func Generate(db db.RWDatabase, trieVersion trie.TrieLayout, rootHash common.Hash, keys []string) (
```

Comment:
Maybe use constructor syntax? `proof.New()`

Reply:
Done! and I also create an alias for the proof